File: examples/ts/byte-builtins/ByteBuiltins.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';
import { Hash } from '@bsv/sdk';

/**
 * ByteBuiltins at the EXAMPLES layer — the interpreter half of the oracle.
 *
 * `conformance/byte_builtins_execution_test.go` spends this contract's compiled
 * bytes on the go-sdk consensus interpreter. This file reads the same source
 * through the ANF interpreter instead, so the two halves are independent: a
 * codegen bug moves the Go test, a lowering-or-interpreter disagreement moves
 * this one, and only a bug in the SOURCE SEMANTICS moves both.
 *
 * The boundaries mirror the Go test's, because that is where the value is:
 * split at 0 and at len, int2str's sign-magnitude encoding, reverseBytes on
 * empty and odd-length input. Each group carries a negative so a builtin that
 * ignored its arguments could not pass.
 */
const __dirname = dirname(fileURLToPath(import.meta.url));
const FILE = 'ByteBuiltins.runar.ts';
const source = readFileSync(join(__dirname, FILE), 'utf8');

const PREIMAGE = 'runar byte-builtins fixture';
const DIGEST = Buffer.from(
  Hash.sha256(Array.from(Buffer.from(PREIMAGE, 'utf8'))),
).toString('hex');

/** A contract with the SHA-256 digest of PREIMAGE baked in. */
function contract() {
  return TestContract.fromSource(source, { expectedDigest: DIGEST }, FILE);
}

const call = (method: string, args: Record<string, unknown>) =>
  contract().call(method, args);

describe('ByteBuiltins — split', () => {
  // `split` binds the RIGHT half. Index 0 and index == len are the two ends of
  // the legal range and the only places an off-by-one can show.
  it.each([
    ['index 0 binds the whole string', 'aabbccdd', 0n, 'aabbccdd'],
    ['index == len binds the empty string', 'aabbccdd', 4n, ''],
    ['a mid split', 'aabbccdd', 2n, 'ccdd'],
    ['the empty string at index 0', '', 0n, ''],
  ])('accepts %s', (_name, data, idx, expectedTail) => {
    const r = call('checkSplit', { data, idx, expectedTail });
    expect(r.success, r.error).toBe(true);
  });

  it('rejects the wrong tail', () => {
    expect(call('checkSplit', { data: 'aabbccdd', idx: 2n, expectedTail: 'ccde' }).success)
      .toBe(false);
  });

  it('rejects the LEFT half — the binding is the right one', () => {
    expect(call('checkSplit', { data: 'aabbccdd', idx: 2n, expectedTail: 'aabb' }).success)
      .toBe(false);
  });
});

describe('ByteBuiltins — int2str', () => {
  // OP_NUM2BIN is fixed-width little-endian SIGN-MAGNITUDE: the sign lives in
  // the top byte, not as a two's complement. Both wrong encodings are named
  // below so a regression says which one came back.
  it.each([
    ['zero into 4 bytes', 0n, 4n, '00000000'],
    ['one into 4 bytes', 1n, 4n, '01000000'],
    ['minus one into 4 bytes', -1n, 4n, '01000080'],
    ['minus five into 4 bytes', -5n, 4n, '05000080'],
    ['255, which needs its own sign byte', 255n, 2n, 'ff00'],
    ['zero into zero width', 0n, 0n, ''],
  ])('accepts %s', (_name, value, width, expected) => {
    const r = call('checkInt2Str', { value, width, expected });
    expect(r.success, r.error).toBe(true);
  });

  it('rejects a two’s-complement encoding of -1', () => {
    expect(call('checkInt2Str', { value: -1n, width: 4n, expected: 'ffffffff' }).success)
      .toBe(false);
  });

  it('rejects a big-endian encoding of 1', () => {
    expect(call('checkInt2Str', { value: 1n, width: 4n, expected: '00000001' }).success)
      .toBe(false);
  });

  it('refuses a width too small to hold the value', () => {
    // NUM2BIN must ABORT rather than truncate — truncation is how a value
    // silently becomes a different number on chain.
    expect(call('checkInt2Str', { value: 255n, width: 1n, expected: 'ff' }).success)
      .toBe(false);
  });
});

describe('ByteBuiltins — reverseBytes', () => {
  const rev = (hex: string) =>
    Buffer.from(Buffer.from(hex, 'hex')).reverse().toString('hex');

  it.each(['', 'aa', 'aabb', 'aabbcc', '0011223344556677889900aabbccddeeff'])(
    'reverses %s',
    (data) => {
      const r = call('checkReverse', { data, expected: rev(data) });
      expect(r.success, r.error).toBe(true);
    },
  );

  it('reverses a 520-byte input — the maximum BSV stack element', () => {
    // The lowering unrolls exactly 520 iterations; this is the only input for
    // which the LAST of them does real work.
    const data = Buffer.from(Array.from({ length: 520 }, (_, i) => i % 256)).toString('hex');
    const r = call('checkReverse', { data, expected: rev(data) });
    expect(r.success, r.error).toBe(true);
  });

  it('rejects the identity — a no-op reverse must not pass', () => {
    expect(call('checkReverse', { data: 'aabbcc', expected: 'aabbcc' }).success).toBe(false);
  });

  it('rejects a 16-bit word swap', () => {
    expect(call('checkReverse', { data: '01020304', expected: '03040102' }).success).toBe(false);
  });
});

describe('ByteBuiltins — sha256', () => {
  it('accepts the preimage of the baked digest', () => {
    const r = call('checkSha256', { preimage: Buffer.from(PREIMAGE, 'utf8').toString('hex') });
    expect(r.success, r.error).toBe(true);
  });

  it('rejects the DIGEST itself', () => {
    // This is the shape of the Go-surface `runar.Sha256` defect: when the hash
    // call is lowered to an identity binding the contract degenerates to
    // `preimage == storedDigest`, and the public digest becomes the key.
    expect(call('checkSha256', { preimage: DIGEST }).success).toBe(false);
  });

  it('rejects an unrelated preimage', () => {
    expect(call('checkSha256', { preimage: '00' }).success).toBe(false);
  });
});

File: examples/go/byte-builtins/ByteBuiltins_test.go
package contract

import (
	"encoding/hex"
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

// ByteBuiltins.runar.go carried `//go:build ignore` from the day it was written
// until `Ripemd160Hash` reached the seven `.runar.go` type tables, so nothing
// here ever ran: the contract type was not constructible from a test binary and
// the port got only the Rúnar half of what a `.runar.go` file is for.
//
// The two digests below are FIPS/RIPEMD known-answer vectors, not values this
// repo computed. That is deliberate: the SDK mock is the thing under test on
// this side of the fence, so an "expected" value it produced itself would agree
// with any bug it has. `conformance/byte_builtins_execution_test.go` spends the
// compiled script for the same two builtins on the go-sdk consensus
// interpreter, against digests from crypto/sha256 and golang.org/x/crypto —
// three independent implementations, none grading its own output.

// sha256OfABC is SHA-256("abc"), FIPS 180-4 appendix B.1.
const sha256OfABC = "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad"

// ripemd160OfABC is RIPEMD-160("abc") from the algorithm's published test
// vectors.
const ripemd160OfABC = "8eb208f7e05d987a9b044a8e98c6b087f15a0bfc"

// ripemd160OfEmpty is RIPEMD-160(""), the other published vector. An
// implementation that returned its input would return the empty string here,
// so this row is what makes the identity-conversion failure visible.
const ripemd160OfEmpty = "9c1185a5c5e9fc54612808977ee8f548b2258d31"

func mustHex(t *testing.T, s string) runar.ByteString {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("bad hex %q: %v", s, err)
	}
	return runar.ByteString(b)
}

// newByteBuiltins bakes the two digests of `preimage` into the contract the way
// the locking script would.
func newByteBuiltins(preimage runar.ByteString) *ByteBuiltins {
	return &ByteBuiltins{
		ExpectedDigest: runar.Sha256(preimage),
		ExpectedRipemd: runar.Ripemd160(preimage),
	}
}

// mustAccept runs fn and fails if the contract refuses. runar.Assert panics on
// a false condition, mirroring OP_VERIFY aborting the script, so "the contract
// refuses" and "fn panics" are the same event.
func mustAccept(t *testing.T, what string, fn func()) {
	t.Helper()
	defer func() {
		if r := recover(); r != nil {
			t.Fatalf("%s: the contract REFUSED (%v)", what, r)
		}
	}()
	fn()
}

// mustRefuse is the other half: without it every acceptance above would pass on
// a contract whose assertion had been deleted.
func mustRefuse(t *testing.T, what string, fn func()) {
	t.Helper()
	defer func() {
		if recover() == nil {
			t.Fatalf("%s: the contract ACCEPTED where it must refuse", what)
		}
	}()
	fn()
}

func TestByteBuiltins_Split(t *testing.T) {
	c := newByteBuiltins(runar.ByteString("x"))
	data := mustHex(t, "aabbccdd")

	// `split` binds the RIGHT half. 0 and len are the ends of the legal range
	// and the only places an off-by-one shows.
	for _, row := range []struct {
		name string
		idx  int64
		tail string
	}{
		{"index 0 binds the whole string", 0, "aabbccdd"},
		{"a mid split", 2, "ccdd"},
		{"index == len binds the empty string", 4, ""},
	} {
		mustAccept(t, row.name, func() {
			c.CheckSplit(data, row.idx, mustHex(t, row.tail))
		})
	}

	mustRefuse(t, "split at 2 is not the whole string", func() {
		c.CheckSplit(data, 2, data)
	})
}

func TestByteBuiltins_Int2Str(t *testing.T) {
	c := newByteBuiltins(runar.ByteString("x"))

	// Fixed-width little-endian sign-magnitude, which is Script's number
	// encoding and not Go's.
	mustAccept(t, "1 in 4 bytes", func() {
		c.CheckInt2Str(1, 4, mustHex(t, "01000000"))
	})
	mustAccept(t, "-1 in 4 bytes sets the sign bit of the last byte", func() {
		c.CheckInt2Str(-1, 4, mustHex(t, "01000080"))
	})
	mustRefuse(t, "-1 is not encoded as +1", func() {
		c.CheckInt2Str(-1, 4, mustHex(t, "01000000"))
	})
}

func TestByteBuiltins_Reverse(t *testing.T) {
	c := newByteBuiltins(runar.ByteString("x"))

	mustAccept(t, "odd length", func() {
		c.CheckReverse(mustHex(t, "aabbcc"), mustHex(t, "ccbbaa"))
	})
	mustAccept(t, "empty", func() {
		c.CheckReverse(runar.ByteString(""), runar.ByteString(""))
	})
	mustRefuse(t, "a non-palindrome is not its own reverse", func() {
		c.CheckReverse(mustHex(t, "aabbcc"), mustHex(t, "aabbcc"))
	})
}

// TestByteBuiltins_Sha256 uses the `Sha256Hash` alias spelling, which is what
// CheckSha256 calls and which only the Go surface parser maps.
func TestByteBuiltins_Sha256(t *testing.T) {
	c := &ByteBuiltins{ExpectedDigest: mustHex(t, sha256OfABC)}

	mustAccept(t, "the FIPS 180-4 vector for \"abc\"", func() {
		c.CheckSha256(runar.ByteString("abc"))
	})
	mustRefuse(t, "a different preimage", func() {
		c.CheckSha256(runar.ByteString("abd"))
	})
	// The identity-conversion failure: if Sha256Hash returned its input, a
	// contract whose baked digest IS the digest would accept the digest as the
	// preimage, and the publicly readable locking script would be the key.
	mustRefuse(t, "the baked digest pushed as the preimage", func() {
		c.CheckSha256(runar.ByteString(c.ExpectedDigest))
	})
}

// TestByteBuiltins_Ripemd is the half this port exists for. `runar.Ripemd160`
// is a FUNCTION and `runar.Ripemd160Hash` is the digest TYPE; the field above
// could not be spelled at all until the type name reached the parsers.
func TestByteBuiltins_Ripemd(t *testing.T) {
	c := &ByteBuiltins{ExpectedRipemd: mustHex(t, ripemd160OfABC)}

	mustAccept(t, "the published vector for \"abc\"", func() {
		c.CheckRipemd(runar.ByteString("abc"))
	})
	mustRefuse(t, "a different preimage", func() {
		c.CheckRipemd(runar.ByteString("abd"))
	})
	mustRefuse(t, "the baked digest pushed as the preimage", func() {
		c.CheckRipemd(runar.ByteString(c.ExpectedRipemd))
	})

	// The empty-string vector, where an identity conversion is unmistakable.
	empty := &ByteBuiltins{ExpectedRipemd: mustHex(t, ripemd160OfEmpty)}
	mustAccept(t, "the published vector for the empty string", func() {
		empty.CheckRipemd(runar.ByteString(""))
	})
}

func TestByteBuiltins_Compile(t *testing.T) {
	if err := runar.CompileCheck("ByteBuiltins.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}
